hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestClientScannerTimeouts.java

```diff
@@ -0,0 +1,465 @@
+    conf.setInt("hbase.regionserver.msginterval", 3 * 10000);
+    conf.setInt(HConstants.HBASE_RPC_TIMEOUT_KEY, rpcTimeout);
+    conf.setStrings(HConstants.REGION_SERVER_IMPL, RegionServerWithScanTimeout.class.getName());
+    conf.setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER, CLIENT_RETRIES_NUMBER);
+    conf.setInt(HConstants.HBASE_CLIENT_PAUSE, 1000);
+    TEST_UTIL.startMiniCluster(1);
+
+    conf.setInt(HConstants.HBASE_CLIENT_SCANNER_TIMEOUT_PERIOD, scanTimeout);
+    conf.setInt(HBASE_CLIENT_META_READ_RPC_TIMEOUT_KEY, metaScanTimeout);
+    conf.setInt(HBASE_CLIENT_META_SCANNER_TIMEOUT, metaScanTimeout);
+    ASYNC_CONN = ConnectionFactory.createAsyncConnection(conf).get();
+    CONN = ASYNC_CONN.toConnection();
+  }
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    CONN.close();
+    ASYNC_CONN.close();
+    TEST_UTIL.shutdownMiniCluster();
+  }
+
+  public void setup(boolean isSystemTable) throws IOException {
+    RSRpcServicesWithScanTimeout.reset();
+
+    String nameAsString = name.getMethodName();
+    if (isSystemTable) {
+      nameAsString = NamespaceDescriptor.SYSTEM_NAMESPACE_NAME_STR + ":" + nameAsString;
+    }
+    final TableName tableName = TableName.valueOf(nameAsString);
+
+    TEST_UTIL.createTable(tableName, FAMILY);
+    table = CONN.getTable(tableName);
+    asyncTable = ASYNC_CONN.getTable(tableName);
+    putToTable(table, ROW0);
+    putToTable(table, ROW1);
+    putToTable(table, ROW2);
+    putToTable(table, ROW3);
+    LOG.info("Wrote our four values");
+
+    table.getRegionLocator().getAllRegionLocations();
+
+    // reset again incase the creation/population caused anything to trigger
+    RSRpcServicesWithScanTimeout.reset();
+  }
+
+  private void expectRow(byte[] expected, Result result) {
+    assertTrue("Expected row: " + Bytes.toString(expected),
+      Bytes.equals(expected, result.getRow()));
+  }
+
+  private void expectNumTries(int expected) {
+    assertEquals(
+      "Expected tryNumber=" + expected + ", actual=" + RSRpcServicesWithScanTimeout.tryNumber,
+      expected, RSRpcServicesWithScanTimeout.tryNumber);
+    // reset for next
+    RSRpcServicesWithScanTimeout.tryNumber = 0;
+  }
+
+  /**
+   * verify that we don't miss any data when encountering an OutOfOrderScannerNextException.
+   * Typically, the only way to naturally trigger this is if a client-side timeout causes an
+   * erroneous next() call. This is relatively hard to do these days because the server attempts to
+   * always return before the timeout. In this test we force the server to throw this exception, so
+   * that we can test the retry logic appropriately.
+   */
+  @Test
+  public void testRetryOutOfOrderScannerNextException() throws IOException {
+    expectRetryOutOfOrderScannerNext(this::getScanner);
+  }
+
+  /**
+   * AsyncTable version of above
+   */
+  @Test
+  public void testRetryOutOfOrderScannerNextExceptionAsync() throws IOException {
```

Comment:
even though Table and AsyncTable are now backed by the same underlying implementation, they are configured differently (See TableBuilderBase vs AsyncTableBuilderBase). So until we unify that, I am running all of these tests separately against Table and AsyncTable.